cvc.1
.\" Copyright (c) 2004-2005 rPath, Inc.
.TH CVC 1 "12 October 2005" "rPath, Inc."
.SH NAME
cvc \- Conary source management
.SH SYNOPSIS
.B cvc \fIcommand \fR[\fI<args>\fR]
.SH DESCRIPTION
\fBConary\fR is a software configuration manager that uses distributed,
network-based, versioned repositories rather than disparate package
files as its canonical source of data.  Conary uses the repository
versioning to intelligently merge configuration files, file ownership
and permissions, and so forth.  The \fIC\fPonary \fIv\fPersion
\fIc\fPontrol program, called \fBcvc\fP, handles build and source
management, and is responsible for all changes to the repository.
.SH REFERENCE
cvc has several common command-line arguments, all of which modify 
configuration entries and are documented with other configuration
entries.  The rest of its command-line arguments are specific to
which cvc \fBcommand\fP you are running, and are documented with
the commands.
.SS "Building Components and Packages"
A source component is checked out into a \fBsource directory\fP
with the \fBcheckout\fP command, or a source component and
source directory are created with the \fBcvc newpkg\fP subcommand.
Then the contents of the source directory are modified with other
subcommands, and the changes to the source component are then
committed to the repository with the \fBcvc commit\fP subcommand.
Only the \fBcvc commit\fP and \fBcvc sign\fP commands modify the repository; 
the rest of the commands either query the repository for information about a 
source package or schedule a change to be applied to the repository by the next 
\fBcvc commit\fP command.
.PP
The default branch on which each of these commands operates
is specified by the \fBbuildLabel\fP configuration entry
(see below).
.PP
The following \fBcommands\fP to \fBcvc\fP are used to create,
modify, and cook source components:
.TP 4
.B cook \fI[\fR\-\-prep\fI] [\fR--macros file\fI] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-resume [<linenums>|policy]] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-flavor <flavor>] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-macro '<macro> <string>'] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-no-clean] [\fR\-\-unknown-flags] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-no-deps] \e
.PD 0
.TP
.B \ \ \ \ \ [\fR\-\-signature-key <fingerprint>|None] \e
.PD 0
.TP
.B \ \ \ \ \ \fI<file>\fP.recipe\fI|<component>[=<version>]...
.PD

Build a package.  If \fB\-\-prep\fP is specified, the souce code is
unpacked, but is not built. if \fB\-\-macros\fP is specified, it
names a file containing macro definitions to apply to the recipe
macro set.  
.IP
If \fB\-\-resume <linenums>\fP is specified, 
conary will cook the specified line numbers of the recipe.  The format of
<linenums> is any number of pairs of n or n:m 
separated by commas, where n and m are line numbers in the recipe.  
If the initial line number in a pair is left off, leaving :m, 
conary will begin cooking from the beginning of the file.  
If the final line number is left off, leaving n:, conary will cook
until the end 
of the file, run all the policy, and create a changeset.  A \-\-resume n without
a colon is equivalent to n:n, if there are other ranges given to \-\-resume,
or n: if it is the only argument given to \-\-resume.  All resume line 
numbers must be provided in increasing order.

If \fB\-\-resume\fP is specified with no parameter,
the recipe will be recooked from the line of last failure.  
If \fB\-\-resume policy\fP is specified, the policy will be rechecked 
from the last build; this only works if the destdir still exists,
either because of a policy error or because the \fB\-\-no-clean\fP option
was provided.

Example: cvc cook \-\-resume 5:10,15,35:

Note that the \fB\-\-resume\fP option may
be used only with local recipe files, not source components being
built directly from a repository.
.IP 
The \-\-flavor option can be used to override the default Use flags, 
package-specific flags, and arch Flags  for this instantiaton.   
The format is exactly as it is shown in conary config flavors.  The 
flavor specified here overrides flavors specified in the buildFlavor
config option.
.IP
The \-\-macro option assigns the given string value to <macro>.  
The value given overrides any attempt to reset the value later in the 
recipe.  The macro and its boolean value must be enclosed in quotes.
.IP 
The \-\-no-clean option does not clean up the build directories
after the build is complete, even if the build is successful.
This can be useful for rerunning policy for a complete build using 
\-\-resume, or for examining the completed build directories.
.IP 
The \-\-no-deps option prevents conary from raising an error 
if a component or package listed as a build requirement for the package
is not installed on the build system.
.IP
If \-\-signature\-key is specified, the OpenPGP key in the local keyring with 
that fingerprint will be used to create trove signatures for all troves created 
during the cook. The repository must recognize the OpenPGP key that made the 
trove signature or the respective troves will be rejected. If 
\-\-signature\-key None is specified, generation of trove signatures will be 
disabled.
.IP

If the thing to be cooked ends in \fB.recipe\fP, then
it must be a recipe file and the package will be cooked from the
filesystem.  Otherwise, it must name a repository component and
the package will be cooked from the repository.  If a version is 
specified, it must be a source version for that component.
.TP 4
.B add \fIfile...\fP
Adds all the files listed on the command line to the source
component.
.TP
.B annotate \fI<file>\fP
Show the version, date, and author of each of the lines in \fI<file>.
\fI<file> must be a file in a conary source dir.
.TP
.B branch \fInewbranch [\fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP]]*
Creates a new branch (identified by the label \fInewbranch\fP) in the
repository, relative to the branch specified in \fIversion\fP.
By default, if a binary trove is given, the binary and its source are
branched.  If \fB-\-source-only\fR is specified, the source
trove for each binary trove is branched, instead of the binary trove.
If \fB-\-binary-only\fR is specified, no source troves are branched.
.TP
.B checkout [\-\-dir \fI<dir>\fP] \fI<trove\fP[=\fIversion\fP]>\fP
Check the \fI<trove>\fB:source\fR component out of the repository
and put it in directory \fI<dir>\fP if specified, and directory
\fI<trove>\fP otherwise.  Fetches the most recent version
specified by \fBbuildLabel\fP unless \fI<version>\fP is specified.
.TP
.B commit [\-\-message \fI<message>\fP]\fP
Different from \fBconary commit\fP, \fBcvc commit\fP
commits all the changes in the source directory to the repository. 
It will ask for a changelog message unless one is passed on the
command line with \fB\-\-message\fP.  
.TP
.B context [name]\fP
When not passed a name, displays information about the current context.  
See \fIman conary\fP for more information on contexts.
.IP
When passed a name, sets the context to <name> for the current directory.
That context will then be used for all conary
operations in the directory.  The commands \fBcvc checkout\fP and 
\fBcvc newpkg\fP will transfer the current context into their newly 
created directories.
.B describe \fI<xml file>\fP
Update the metadata of the source trove in the current source directory
based on the contents of \fI<xml file>\fP.
.TP
.B diff
Show (in a slightly extended unified diff format) the changes that
have been made in the current source directory since the last
\fBcvc commit\fP
(or, if no commit, since the source component was checked out).
.TP
.B log [\fI<branch>\fP]
Prints the log messages for the branch specified by \fBbuildLabel\fP,
or for \fI<branch>\fP if specified.
.TP
.B newpkg \fI<name\fP[=\fIlabel\fP]>\fP
Creates a new package.
.TP
.B merge
Run from a source directory containing a shadow, this merges changes made
on the parent branch since the last \fBshadow\fP or \fBmerge\fP command
into the shadow.
.TP
.B rdiff \fI<name> <oldver> <newver>\fP
This source command operates only on the repository, not on a
source directory.  It creates a diff between two versions of
a source trove from the repository.
.TP
.B remove \fI<filename>...\fP
Unlike \(lq\fBcvs remove\fP\(rq, \fBcvc remove\fP both removes
the file from the filesystem and marks it to be removed from the next
version checked into the repository at the next \fBcvc commit\fP.
.TP
.B rename \fI<oldname> <newname>\fP
Renames the file \fI<oldname>\fP to \fI<newname>\fP on the filesystem,
and marks it to be removed from the repository at the next
\fBcvc commit\fP.
.TP
.B shadow \fInewshadow [\fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP]]*
Creates a new shadow (identified by the label \fInewshadow\fP) in the
repository, relative to the shadow specified in \fIversion\fP.
By default, if a binary trove is given, the binary and its source are
shadowed.  If \fB-\-source-only\fR is specified, the source
trove for each binary trove is shadowed, instead of the binary trove.
If \fB-\-binary-only\fR is specified, no source troves are shadowed.
The shadow command works just like the branch command, except that
it creates a shadow instead of a branch.
.TP
.B sign [\fR\-\-signature\-key <fingerprint>\fB] 
[\fItrovename\fP[=\fIversion\fP][\fI<flavor>\fP]]*
Signs the specified troves in their respective repositories. If 
\fB-\-signature\-key\fR is specified, it will override all signatureKey and 
signatureKeyMap settings specified in config files. The repositories affected 
must recognize the OpenPGP key(s) that made the trove signature(s) or the 
respective signatures will be rejected.
.TP
.B update [\fI<version>\fP]
Updates the current source directory to the latest version, or to
\fI<version>\fP if specified.  Merges changes when possible.
.RE
.\"
.\"
.\"
.SH TROVE SIGNATURES
.TP 4
.B Discussion
.br
cvc is able to generate trove signatures. in all cases <fingerprint> refers to 
the fingerprint of an OpenPGP key from the local keyring. (generally ~/.gnupg). 
When a signature is generated and submitted to a repository, the repository 
verifies the signature. If the signature is bad, or the repository doesn't know 
about the OpenPGP Key that made that signature, the repository will remain 
unachaged. Aka, a signature failure on cook will result in the trove not being 
uploaded to the repository, whereas a failure on sign will not affect the trove 
on the repository, whether or not that trove already had signatures associated 
with it. The OpenPGP key that will be used can be specified on the command line 
or in a config file. see conary(1) for a discussion on the appropriate settings.
.TP 4
.B Fingerprints
.br
Fingerprints can have spaces in them. Valid OpenPGP key ids can also be used in
place of the full fingerprint. For example:
.br
\ \ \ \ AAFF BBCC DDEE 1122
.br
\ \ \ \ 1234 5678 9ABC DEF0 1212 3344 AAFF BBCC DDEE 1122
.br
\ \ \ \ 123456789ABCDEF012123344AAFFBBCCDDEE1122
.br
All refer to the exact same OpenPGP Key
.RE
.\"
.\"
.\"
.SH EXAMPLES
.TP 4
.B Branching
cvc branch conary.example.com@rpl:example-foo foo:source=conary.rpath.com@rpl:devel
.br
This creates a branch in the conary.example.com repository of the
foo:source trove, based on the version of foo in the main rpath repository.
You can now check out this branch and work on it:
.br
cvc checkout --build-label conary.example.com@rpl:example-foo foo
.br
Committing changes to that copy of foo:source will go into the
conary.example.com repository on the rpl:example-foo branch.
.P
More examples coming soon to a man page near you!
.\"
.\"
.\"
.SH BUGS
There are no bugs, only undocumented features.  You can report
undocumented features at http://bugs.rpath.com/
.\"
.\"
.\"
.SH "SEE ALSO"
conary(1)
.br
cvcdesc(1)


NEWS
Changes in 0.71.1:
  o Server Changes 
    * Added iterTroves() call which iterates over large numbers of troves
      much more efficiently than a single getTrove() call would.
    * Split out FileRetriever object to allow file information to be pulled
      from the repository inside of an iterTroves() loop
  o Client Changes
    * Config file options that take a path as a value now support ~ for 
      home directory substitution

Changes in 0.71.0:
  o Code Structure
    * conary now imports all python modules from a toplevel "conary"
      module.  This prevents conary from polluting the module namespace.
  o Client Changes
    * Clone didn't handle shadow version numbers correctly (and could create
      inconsistent version numbers)

Changes in 0.70.5:
  o Client Changes
    * Files changing to config files across distributed repositories now works.
    * The update code uses more consistent use of trove sources, and only
      makes explicit calls to the repository if asked.  This should make it
      possible to create interesting update filters.
    * Clone updated sequences it was iterating over, which is generally
      a bad idea (and caused clone to commit inconsistent troves)
  o Build Changes (cvc)
    * Locally cooked filesets now include file contents, making the
      filesets installable.
    * Fileset cooks now retrieve all of the file objects in a single
      network request per repository.
    * The new NormalizeLibrarySymlinks policy runs the ldconfig program
      in all system library directories.  This ensures that all the
      same symlinks that ldconfig would create when the shlib tag handler
      runs are packaged.  It also warns if ldconfig finds missing files.
    * New argument to r.Run(): "wrapdir" keyword argument behaves much
      like "filewrap" but takes a string argument, which limits the scope of
      %(destdir)s relocation only to the directories under the specified
      wrapdir, which is interpreted relative to %(destdir)s.  Works best
      for applications that install under one single directory, such
      as /opt/<app>
    * Clone, branch, and shadow all take --info now instead of --test
    * ELF files that dlopen() libraries can now be provided with
      synthetic soname dependencies with
      r.Requires('soname: libfoo.so', '/path/to/file')
    * r.Requires now enforces that packages that require a file and
      include that required file must also explicitly provide it. (bz #148)
  o Server Changes 
    * Packages added to the repository are checked to ensure the version and
      flavor of all referenced components are the same as for the package

Changes in 0.70.4:
  o Client Changes
    * The trove that satisfies a dependency that is broken by erase is
      now displayed in the "Troves being removed create unresolved
      dependencies" message.
    * Components are now displayed on the same line as their parent
      package in "conary update" output.
    * A new 'interactive' option has been added to conary configuration. 
      When set to true, conary will display info about clone, branch, 
      update, and erase operations, and then ask before proceding.
  o Build Changes (cvc)
    * The CompilePython action has been fixed to accept macros at the
      beginning of its arguments, fixing a bug new in Conary 0.70.3.
    * The Requires policy can now be given synthetic RPATH elements;
      this is useful when programs are only intended to be run under
      scripts that set LD_LIBRARY_PATH and so do not intrinsically have
      the information they need to find their libraries.
    * Added --test to clone, branch, and shadow commands
    * Clone now supports --skip-build-info for less rigid version checks
      on cloned troves
    * Fixed usage message to better reflect reality
    * Cloning to a branch which already has a version with a compatible
      flavor now works.
    * cpio archive files are now supported for r.addArchive()
  o Repository Changes
    * The repository now serves up stored OpenPGP keys as a "Limited
      Keyserver"; users can retrieve keys, but not search or browse them.
      The keys are available via /getOpenPGPKey?search=KEY_ID. This
      is meant only to allow conary to automatically retrieve OpenPGP
      keys used to sign packages.

Changes in 0.70.3:
  o Client Changes (conary)
    * Conary now works harder to avoid having separate erase/installs,
      instead preferring to link those up into one update when possible.
    * Conary configuration now supports contexts.  Contexts are defined in 
      sections starting with a [<name>] line, and provide contextual 
      configurations for certain variables, defined in the man page.  All 
      configuration options after the [<name>] will be associated with that 
      context, and will override the default configuration when that context 
      is active.  The current context can be selected by using the --context 
      parameter, or by setting the CONARY_CONTEXT environment variable. 
    * 'conary config --show-contexts' will display the available contexts
  o Build Changes (cvc)
    * A local cook of a trove foo will ensure that the changeset created is 
      installable on your local system, by making sure the version number 
      created is unique.
    * The builddir is no longer allowed to appear in ELF RPATHs.
    * The build documentation strings have been significantly updated
      to document the fact that for most strings, a relative path
      is relative to the builddir, but an absolute path is relative
      to the destdir.
    * The ManualConfigure action now sets the standard Configure
      environment.
    * cvc will allow you to cook a trove locally even when you are unable 
      to access the trove's source repository
  * Common Changes:
    * Version closeness was improperly measured for troves on different
      branches when then label structure was identical
  o Repository Changes

conary/cvc.py
# -*- mode: python -*-
#
# Copyright (c) 2004-2005 rPath, Inc.
#
# This program is distributed under the terms of the Common Public License,
# version 1.0. A copy of this license should have been distributed with this
# source file in a file called LICENSE. If it is not present, the license
# is always available at http://www.opensource.org/licenses/cpl.php.
#
# This program is distributed in the hope that it will be useful, but
# without any waranty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the Common Public License for
# full details.
#
"""
Provides the output for the "cvc" subcommands
"""

import os
import sys
import xmlrpclib

from conary import branch
from conary import checkin
from conary import conarycfg
from conary import constants
from conary import deps
from conary import flavorcfg
from conary import updatecmd
from conary import versions
from conary.build import cook, use, signtrove
from conary.lib import log
from conary.lib import openpgpfile
from conary.lib import options
from conary.lib import util
from conary.local import database
from conary.repository import errors, netclient
from conary.repository.netclient import NetworkRepositoryClient

sys.excepthook = util.genExcepthook()

# mix UpdateCallback and CookCallback, since we use both.
class CheckinCallback(updatecmd.UpdateCallback, cook.CookCallback):
    def __init__(self):
        updatecmd.UpdateCallback.__init__(self)
        cook.CookCallback.__init__(self)

def usage(rc = 1):
    print "usage: cvc add <file> [<file2> <file3> ...]"
    print "       cvc annotate <file>"
    print "       cvc branch <newbranch> <trove>[=<version>][[flavor]]"
    print "       cvc checkout [--dir <dir>] [--trust-threshold <int>]"
    print "                    <trove>[=<version>]"
    print "       cvc clone <target-branch> <trove>[=<version>][[flavor]]+ "
    print "                 [--skip-build-info]"
    print "       cvc commit [--message <message>]"
    print '                  [--signature-key "<fingerprint>"]'
    print "       cvc config"
    print "       cvc context [<name>] [--ask]"
    print '       cvc cook [--prep] [--debug-exceptions] [--macros file] '
    print '                [--flavor  "<flavor>"] '
    print '                [--signature-key "<fingerprint>"]'
    print '                [--macro "<macro> <value>"]+ '
    print '                <file.recipe|troveName=<version>>[[flavor]]+'
    print '       cvc describe <xml file>'
    print "       cvc diff"
    print "       cvc log [<branch>]"
    print "       cvc newpkg [--dir <dir>] <name>"
    print "       cvc merge"
    print "       cvc rdiff <name> <oldver> <newver>"
    print "       cvc remove <file> [<file2> <file3> ...]"
    print "       cvc rename <oldfile> <newfile>"
    print "       cvc shadow <newshadow> <trove>[=<version>][[flavor]]"
    print '       cvc sign [--signature-key "<fingerprint>"]'
    print '                [--quiet] <trove>[=version][[flavor]]'
    print "                <trove2>[=version2][[flavor2]]..."
    print "       cvc update <version>"
    print 
    print "branch flags:  --binary-only"
    print "               --info"
    print "               --source-only"
    print
    print "clone flags:   --info"
    print "               --skip-build-info"
    print
    print "commit flags:  --message <msg>"
    print 
    print 'common flags:  --build-label <label>'
    print '               --config-file <path>'
    print '               --config "<item> <value>"'
    print '               --context <context>'
    print '               --install-label <label>'
    print "               --root <root>"
    print '               --signature-key "<fingerprint>"'
    print '               --trust-threshold <int>'
    print ""
    print "cook flags:    --macros"
    print "               --no-clean"
    print '               --unknown-flags'
    print '               --no-deps'
    print '               --flavor  "<flavor>"'
    print '               --macro "<macro> <value>"'
    print "               --prep"
    print "               --resume [policy|<linenums>]"
    print "               --debug-exceptions"
    print "               --quiet"
    print ""
    print "shadow flags:  --binary-only"
    print "               --info"
    print "               --source-only"
    print ""
    print "sign flags:    --quiet"
    print '               --signature-key "<fingerprint>"'
    return rc

def realMain(cfg, argv=sys.argv):
    argDef = {}
    cfgMap = {}

    cfgMap["build-label"] = "buildLabel"
    cfgMap["signature-key"] = "signatureKey"
    cfgMap["trust-threshold"] = "trustThreshold"

    (NO_PARAM,  ONE_PARAM)  = (options.NO_PARAM, options.ONE_PARAM)
    (OPT_PARAM, MULT_PARAM) = (options.OPT_PARAM, options.MULT_PARAM)

    argDef["ask"] = NO_PARAM
    argDef["binary-only"] = NO_PARAM
    argDef["config"] = MULT_PARAM
    argDef["config-file"] = ONE_PARAM
    argDef["context"] = ONE_PARAM
    argDef["debug-exceptions"] = NO_PARAM
    argDef["dir"] = ONE_PARAM
    argDef["flavor"] = ONE_PARAM
    argDef["info"] = NO_PARAM
    argDef["macro"] = MULT_PARAM
    argDef["macros"] = ONE_PARAM
    argDef["message"] = ONE_PARAM
    argDef["no-clean"] = NO_PARAM
    argDef["no-deps"] = NO_PARAM
    argDef["prep"] = NO_PARAM
    argDef["profile"] = NO_PARAM
    argDef["quiet"] = NO_PARAM
    argDef["replace-files"] = NO_PARAM
    argDef["resume"] = OPT_PARAM
    argDef["sha1s"] = NO_PARAM
    argDef["show-passwords"] = NO_PARAM
    argDef["show-contexts"] = NO_PARAM
    argDef["skip-build-info"] = NO_PARAM
    argDef["sources"] = NO_PARAM
    argDef["source-only"] = NO_PARAM
    argDef["tag-script"] = ONE_PARAM
    argDef["tags"] = NO_PARAM
    argDef["unknown-flags"] = NO_PARAM
    argDef["version"] = NO_PARAM

    try:
        argSet, otherArgs = options.processArgs(argDef, cfgMap, cfg, usage,
                                                argv=argv)
    except options.OptionError, e:
        sys.exit(e.val)
    except versions.ParseError, e:
        print >> sys.stderr, e
        sys.exit(1)

    if argSet.has_key('version'):
        print constants.version
        sys.exit(0)

    context = cfg.context
    if os.path.exists('CONARY'):
        state = checkin.ConaryStateFromFile('CONARY')
        if state.hasContext():
            context = state.getContext()

    context = os.environ.get('CONARY_CONTEXT', context)
    context = argSet.pop('context', context)


    if context:
        cfg.setContext(context)

    cfg.initializeFlavors()
    # set the build flavor here, just to set architecture information 
    # which is used when initializing a recipe class
    use.setBuildFlagsFromFlavor(None, cfg.buildFlavor, error=False)
    if 'profile' in argSet:
	del argSet['profile']
	import hotshot
        # XXX note that this profile is currently useless for cook - 
        # hotshot seems to not have the correct frame information about this
        # process.
        # Instead, the conary-cook.prof profile has information about the 
        # forked cook process, which is generally more interesting anyway  
	prof = hotshot.Profile('conary.prof')
	prof.start()
        try:
            sourceCommand(cfg, otherArgs[1:], argSet, profile=True)
        finally:
            prof.stop()
    else:
        sourceCommand(cfg, otherArgs[1:], argSet)

def sourceCommand(cfg, args, argSet, profile=False, callback = None):
    if not callback:
        callback = CheckinCallback()

    if not args:
	return usage()
    elif (args[0] == "add"):
	if len(args) < 2: return usage()
	checkin.addFiles(args[1:])
    elif (args[0] == "checkout") or (args[0] == "co"):
	if argSet.has_key("dir"):
	    dir = argSet['dir']
	    del argSet['dir']
	else:
	    dir = None

	if argSet or (len(args) != 2): return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)

	args = [repos, cfg, dir, args[1], callback]
	checkin.checkout(*args)
    elif (args[0] == "branch" or args[0] == "shadow"):
        makeShadow =  (args[0] == "shadow")
        sourceOnly = argSet.pop('source-only', False)
        binaryOnly = argSet.pop('binary-only', False)
        info = argSet.pop('info', False)

        if argSet: return usage()
        if len(args) < 3: return usage()

	repos = NetworkRepositoryClient(cfg.repositoryMap)
        target = args[1]
        troveSpecs = args[2:]

        branch.branch(repos, cfg, target, troveSpecs, makeShadow = makeShadow, 
                      sourceOnly = sourceOnly, binaryOnly = binaryOnly,
                      info = info)

    elif (args[0] == "commit") or (args[0] == "ci"): # mimic cvs's shortcuts
        level = log.getVerbosity()
        log.setVerbosity(log.INFO)
	message = argSet.get("message", None)
        sourceCheck = True

	if message is not None:
	    del argSet['message']

	if argSet or len(args) != 1: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)

	checkin.commit(repos, cfg, message, callback=callback)
        log.setVerbosity(level)
    elif (args[0] == "config"):
	showPasswords = argSet.pop('show-passwords', False)
	showContexts = argSet.pop('show-contexts', False)
        try:
            prettyPrint = sys.stdout.isatty()
        except AttributeError:
            prettyPrint = False
        cfg.setDisplayOptions(hidePasswords=not showPasswords,
                              showContexts=showContexts,
                              prettyPrint=prettyPrint)
	if argSet: return usage()
	if (len(args) > 2):
	    return usage()
	else:
	    cfg.display()
    elif (args[0] == "clone"):
        if len(args) < 3:
            return usage()

        import clone
        skipBuildInfo = argSet.pop('skip-build-info', False)
        info = argSet.pop('info', False)
        if argSet: return usage()
        clone.CloneTrove(cfg, args[1], args[2:], not skipBuildInfo, info = info)
    elif (args[0] == "diff"):
	if argSet or not args or len(args) > 2: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)

	args[0] = repos
	checkin.diff(*args)
    elif (args[0] == "annotate"):
	if argSet or len(args) != 2: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)
	args[0] = repos
	checkin.annotate(*args)

    elif (args[0] == "log"):
	if argSet or len(args) > 2: return usage()

	repos = NetworkRepositoryClient(cfg.repositoryMap)
	args[0] = repos
	checkin.showLog(*args)
    elif (args[0] == "rdiff"):
	if argSet or len(args) != 4: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)

	checkin.rdiff(repos, cfg.buildLabel,  *args[1:])
    elif (args[0] == "remove") or (args[0] == "rm"):
	if len(args) < 2: return usage()
        for f in args[1:]:
            checkin.removeFile(f)
    elif (args[0] == "rename"):
	if len(args) != 3: return usage()
	checkin.renameFile(args[1], args[2])
    elif (args[0] == "sign"):
        if len(args) <2: return usage()
        if argSet.has_key('quiet'):
            cfg.quiet = True
            del argSet['quiet']
        signtrove.signTroves(cfg, args[1:])
    elif (args[0] == "newpkg"):
        dir = argSet.pop('dir', None)

	if len(args) != 2 or argSet: return usage()
	
	try:
	    repos = NetworkRepositoryClient(cfg.repositoryMap)
	except errors.OpenError:
	    repos = None

	checkin.newTrove(repos, cfg, args[1], dir = dir)
    elif (args[0] == "context"):
        if len(args) > 2:
            return usage()

        ask = argSet.pop('ask', False)
        if len(args) > 1:
            name = args[1]
        else:
            name = None

	checkin.setContext(cfg, name, ask=ask)
    elif (args[0] == "merge"):
	if argSet or not args or len(args) > 1: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)
        
	checkin.merge(repos)
    elif (args[0] == "update") or (args[0] == "up"):
	if argSet or not args or len(args) > 2: return usage()
	repos = NetworkRepositoryClient(cfg.repositoryMap)

	args[0] = repos
        kwargs = {'callback': callback}
	checkin.updateSrc(*args, **kwargs)
    elif (args[0] == "cook"):
        level = log.getVerbosity()
        log.setVerbosity(log.DEBUG)
        macros = {}
        prep = 0
        resume = None
        buildBranch = None
        if argSet.has_key('flavor'):
            buildFlavor = deps.deps.parseFlavor(argSet['flavor'])
            cfg.buildFlavor = deps.deps.overrideFlavor(cfg.buildFlavor,
                                                       buildFlavor)
            del argSet['flavor']
        if argSet.has_key('macro'):
            for macro in argSet['macro']:
                cfg.configLine('macros.' + macro)
            del argSet['macro']

        if argSet.has_key('prep'):
            del argSet['prep']
            prep = 1
        if argSet.has_key('no-deps'):
            del argSet['no-deps']
            ignoreDeps = True
        else:
            ignoreDeps = False

        if argSet.has_key('quiet'):
            cfg.quiet = True
            del argSet['quiet']

        if argSet.has_key('no-clean'):
            del argSet['no-clean']
            cfg.noClean = True
        else:
            cfg.noClean = False
        if argSet.has_key('resume'):
            resume = argSet['resume']
            del argSet['resume']
        if argSet.has_key('unknown-flags'):
            unknownFlags = argSet['unknown-flags']
            del argSet['unknown-flags']
        else:
            unknownFlags = False
        if argSet.has_key('debug-exceptions'):
            del argSet['debug-exceptions']
            cfg.debugRecipeExceptions = True
        if argSet.has_key('macros'):
            argSet['macros']
            f = open(argSet['macros'])
            # XXX sick hack
            macroSrc = "macros =" + f.read()
            exec macroSrc
            del f
            del argSet['macros']

        if argSet: return usage()
        
        cook.cookCommand(cfg, args[1:], prep, macros, resume=resume, 
                         allowUnknownFlags=unknownFlags, ignoreDeps=ignoreDeps,
                         profile=profile)
        log.setVerbosity(level)
    elif (args[0] == "describe"):
        level = log.getVerbosity()
        log.setVerbosity(log.INFO)
        
        xmlSource = args[1]
        state = checkin.ConaryStateFromFile("CONARY").getSourceState()
        troveName = state.getName()
        troveBranch = state.getVersion().branch()
       
        log.info("describing trove %s with %s", troveName, xmlSource)
        xmlFile = open(xmlSource)
        xml = xmlFile.read()

        repos = NetworkRepositoryClient(cfg.repositoryMap)
        repos.updateMetadataFromXML(troveName, troveBranch, xml)
        log.setVerbosity(level)
    elif (args[0] == "usage"):	
        return usage(rc = 0)
    else:
	return usage()

    return 0

def main(argv=sys.argv):
    try:
        if '--skip-default-config' in argv:
            argv = argv[:]
            argv.remove('--skip-default-config')
            cfg = conarycfg.ConaryConfiguration(False)
        else:
            cfg = conarycfg.ConaryConfiguration()
        # reset the excepthook (using cfg values for exception settings)
        sys.excepthook = util.genExcepthook(cfg.dumpStackOnError)
	return realMain(cfg, argv)
    except conarycfg.ConaryCfgError, e:
        log.error(str(e))
        sys.exit(1)
    except xmlrpclib.ProtocolError, e:
        if e.errcode == 403:
            print >> sys.stderr, \
                "remote server denied permission for the requested operation"
        else:
            raise
    except errors.UnknownException, e:
        print >> sys.stderr, \
            "An unknown exception occured on the repository server:"
        print >> sys.stderr, "\t%s" % str(e)
    except errors.RepositoryError, e:
        print >> sys.stderr, str(e)
    except database.OpenError, e:
        print >> sys.stderr, str(e)
    except checkin.ConaryStateError, e:
        print >> sys.stderr, str(e)
    except openpgpfile.KeyNotFound, e:
        print >> sys.stderr, str(e)
    except KeyboardInterrupt, e:
        pass
    return 1


if __name__ == "__main__":
    sys.exit(main())

